oauth-client/src/main/java/io/streamnative/pulsar/handlers/kop/security/oauth/ClientCredentialsFlow.java

```diff
@@ -0,0 +1,171 @@
+/**
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.streamnative.pulsar.handlers.kop.security.oauth;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.ObjectReader;
+import com.google.common.annotations.VisibleForTesting;
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.UnsupportedEncodingException;
+import java.net.URI;
+import java.net.URL;
+import java.net.URLConnection;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Collectors;
+import lombok.Getter;
+import org.asynchttpclient.AsyncHttpClient;
+import org.asynchttpclient.DefaultAsyncHttpClient;
+import org.asynchttpclient.DefaultAsyncHttpClientConfig;
+import org.asynchttpclient.Response;
+
+/**
+ * The OAuth 2.0 client credential flow.
+ */
+public class ClientCredentialsFlow implements Closeable {
+
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+    private static final ObjectReader METADATA_READER = OBJECT_MAPPER.readerFor(Metadata.class);
+    private static final ObjectReader CLIENT_INFO_READER = OBJECT_MAPPER.readerFor(ClientInfo.class);
+    private static final ObjectReader TOKEN_RESULT_READER = OBJECT_MAPPER.readerFor(OAuthBearerTokenImpl.class);
+    private static final ObjectReader TOKEN_ERROR_READER = OBJECT_MAPPER.readerFor(TokenError.class);
+
+    private final Duration connectTimeout = Duration.ofSeconds(10);
+    private final Duration readTimeout = Duration.ofSeconds(30);
+    private final ClientConfig clientConfig;
+    private final AsyncHttpClient httpClient;
+
+    public ClientCredentialsFlow(ClientConfig clientConfig) {
+        this.clientConfig = clientConfig;
+        this.httpClient = new DefaultAsyncHttpClient(new DefaultAsyncHttpClientConfig.Builder()
+                .setFollowRedirect(true)
+                .setConnectTimeout((int) connectTimeout.toMillis())
+                .setReadTimeout((int) readTimeout.toMillis())
+                .build());
+    }
+
+    public OAuthBearerTokenImpl authenticate() throws IOException {
+        final String tokenEndPoint = findAuthorizationServer().getTokenEndPoint();
```

Comment:
Every time we call the authenticate method, we will try to get the token endpoint, I think we need to cache it. Since the token endpoint will not be changed frequently right?

Mabey we can add a config to configure the token end point, so the user can manually specify the token end point.

Reply:
This workflow is the same as the previous implementation.

```java
    private void handleCallback(OAuthBearerTokenCallback callback) throws IOException {
        if (callback.token() != null) {
            throw new IllegalArgumentException("Callback had a token already");
        }
        try (Authentication authentication = AuthenticationFactoryOAuth2.clientCredentials( // 1. find the token endpoint
                clientConfig.getIssuerUrl(),
                clientConfig.getCredentialsUrl(),
                clientConfig.getAudience()
        )) {
            authentication.start(); // 2. get the access token
```

This should not be a frequent work. I think caching the token endpoint is a premature optimization.